solr/core/src/java/org/apache/solr/handler/admin/api/ReplaceNodeAPI.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.handler.admin.CollectionsHandler;
+import org.apache.solr.jersey.JacksonReflectMapWriter;
+import org.apache.solr.jersey.PermissionName;
+import org.apache.solr.jersey.SolrJerseyResponse;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+
+/**
+ * V2 API for recreating replicas in one node (the source) on another node(s) (the target).
+ *
+ * <p>This API is analogous to the v1 /admin/collections?action=REPLACENODE command.
+ */
+@Path("cluster/nodes/{sourceNodeName}/commands/replace/")
+public class ReplaceNodeAPI extends AdminAPIBase {
+
+  @Inject
+  public ReplaceNodeAPI(
+      CoreContainer coreContainer,
+      SolrQueryRequest solrQueryRequest,
+      SolrQueryResponse solrQueryResponse) {
+    super(coreContainer, solrQueryRequest, solrQueryResponse);
+  }
+
+  @POST
+  @Produces({"application/json", "application/xml", BINARY_CONTENT_TYPE_V2})
+  @PermissionName(COLL_EDIT_PERM)
+  public SolrJerseyResponse replaceNode(
+      @Parameter(description = "The name of the node to be replaced.", required = true)
+          @PathParam("sourceNodeName")
+          String nodeName,
+      @RequestBody(description = "Contains user provided parameters", required = true)
+          ReplaceNodeRequestBody requestBody)
+      throws Exception {
+    final SolrJerseyResponse response = instantiateJerseyResponse(SolrJerseyResponse.class);
+    final CoreContainer coreContainer = fetchAndValidateZooKeeperAwareCoreContainer();
+    /** TODO Record node for log and tracing */
+    final ZkNodeProps remoteMessage = createRemoteMessage(nodeName, requestBody);
+    final SolrResponse remoteResponse =
+        CollectionsHandler.submitCollectionApiCommand(
+            coreContainer,
+            coreContainer.getDistributedCollectionCommandRunner(),
+            remoteMessage,
+            CollectionParams.CollectionAction.REPLACENODE,
+            DEFAULT_COLLECTION_OP_TIMEOUT);
+    if (remoteResponse.getException() != null) {
+      throw remoteResponse.getException();
+    }
+
+    disableResponseCaching();
+    return response;
+  }
+
+  public ZkNodeProps createRemoteMessage(String nodeName, ReplaceNodeRequestBody requestBody) {
+    final Map<String, Object> remoteMessage = new HashMap<>();
+    remoteMessage.put(SOURCE_NODE, nodeName);
+    remoteMessage.put(TARGET_NODE, requestBody.targetNodeName);
+    remoteMessage.put(WAIT_FOR_FINAL_STATE, requestBody.waitForFinalState);
+    remoteMessage.put(QUEUE_OPERATION, CollectionAction.REPLACENODE.toLower());
+
+    return new ZkNodeProps(remoteMessage);
+  }
+
+  public static class ReplaceNodeRequestBody implements JacksonReflectMapWriter {
+
+    public ReplaceNodeRequestBody() {}
+
+    public ReplaceNodeRequestBody(String targetNodeName, Boolean waitForFinalState) {
+      this.targetNodeName = targetNodeName;
+      this.waitForFinalState = waitForFinalState;
+    }
+
```

Comment:
[-1] v1 REPLACENODE supports the "async" parameter, so we should probably add that to the RequestBody as well.

(`CollectionsHandler.submitCollectionApiCommand`, which you use above, already has logic built around the parameter, so all we should have to do is add it to the RequestBody here, and then make sure it gets added to the message in `createRemoteMessage` above.)

Reply:
Done